### PR TITLE
Added null annotations to ExpiryCache

### PR DESCRIPTION
**Pull request by cweitkamp**
- Added null annotations to `ExpiryCache` and `ExpiryCacheMap`
- Add constructors with a Duration

Also-by: Martin van Wingerden martin@martinvw.nl
Signed-off-by: Christoph Weitkamp <github@christophweitkamp.de>

### CODE REVIEW

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCache.java [hunk 1]
```diff
@@ -47,15 +49,12 @@ public ExpiringCache(long expiry, Supplier<V> action) {
```
Comment (martinvw) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCache.java]: Imho you should remove this `null` check now

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCacheMap.java [hunk 2]
```diff
@@ -35,6 +36,7 @@
  * @param <K> the type of the key
  * @param <V> the type of the value
  */
+@NonNullByDefault
```
Comment (martinvw) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCacheMap.java]: Then please remove the `null` checks from the constructor and also remove the tests with the null values

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCache.java [hunk 3]
```diff
@@ -28,10 +29,11 @@
  *
  * @param <V> the type of the value
  */
+@NonNullByDefault
 public class ExpiringCache<V> {
     private final long expiry;
     private final Supplier<V> action;
-    private SoftReference<V> value;
+    private SoftReference<V> value = new SoftReference<>(null);
```
Comment (cweitkamp) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCache.java]: @martinvw Do I have to add the `@Nullable` annotation to the type parameter too?

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCache.java (hunk 1 above)
Comment (cweitkamp) [reply bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCache.java]: Done.

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCacheMap.java (hunk 2 above)
Comment (cweitkamp) [reply bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCacheMap.java]: Done.

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCache.java (hunk 3 above)
Comment (martinvw) [reply bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCache.java]: Yes, because AFAIK that was the problem @Hilbrand was facing because he could not pass a nullable supplier to the Cache

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCacheMap.java [hunk 4]
```diff
@@ -208,7 +210,7 @@ public V get(K key) {
      *
      * @return the collection of all values
      */
-    public synchronized Collection<V> values() {
+    public synchronized Collection<@Nullable V> values() {
         final Collection<V> values = new LinkedList<>();
```
Comment (martinvw) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCacheMap.java]: This seems like a mismatch, maybe you should not include the `@Nullable` here?

Comment (cweitkamp) [reply bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCacheMap.java]: But then we have to check if the value we add to the Collection is not `null`.

**Comment by martinvw** [comment]
see https://github.com/cweitkamp/smarthome/pull/1

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCacheMap.java (hunk 4 above)
Comment (martinvw) [reply bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCacheMap.java]: Yes, so then 214 has to be changed as well

Comment (martinvw) [reply bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCacheMap.java]: I did that indeed, so a LinkedList can indeed contain `null` so its all fine.

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCacheMap.java [hunk 5]
```diff
@@ -114,7 +108,7 @@ public void putIfAbsent(K key, ExpiringCache<V> item) {
      * @param action the action for the item to be associated with the specified key to retrieve/calculate the value
      * @return the (cached) value for the specified key
      */
-    public V putIfAbsentAndGet(K key, Supplier<V> action) {
+    public @Nullable V putIfAbsentAndGet(K key, Supplier<V> action) {
```
Comment (martinvw) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCacheMap.java]: It would be nice in hindsight if this `putIfAbsentAndGet` would be Nullable depending on its input :-)

Comment (cweitkamp) [reply bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCacheMap.java]: Yes, and in [line 95](https://github.com/eclipse/smarthome/pull/6482/files#diff-4a40be874170b23672abeb4314145831R95) and [line 124](https://github.com/eclipse/smarthome/pull/6482/files#diff-4a40be874170b23672abeb4314145831R124) too. Will you create a PR?

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCacheAsync.java [hunk 6]
```diff
@@ -38,14 +39,24 @@
...
      * Create a new instance.
      *
-     * @param expiry the duration in milliseconds for how long the value stays valid. Must be greater than 0.
+     * @param expiry the duration in milliseconds for how long the value stays valid. Must be positive.
      * @throws IllegalArgumentException For an expire value <=0.
      */
-    public ExpiringCacheAsync(long expiry) throws IllegalArgumentException {
-        if (expiry <= 0) {
+    public ExpiringCacheAsync(Duration expiry) {
+        if (expiry.isNegative() || expiry.isZero()) {
```
Comment (cweitkamp) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCacheAsync.java]: Shall we add this check to the `ExpiringCache` constructor as well?

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCacheMap.java (hunk 5 above)
Comment (martinvw) [reply bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCacheMap.java]: More thoughts :-)

Given its non-synchronized nature it would be lying someone else could have invalidated the key or the complete cache. I will commit a null check for your code :-)

Comment (martinvw) [reply bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCacheMap.java]: Not needed anymore I see

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCacheAsync.java (hunk 6 above)
Comment (martinvw) [reply bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCacheAsync.java]: Its unlikely that someone constructs it like, but its a framework so th additional robustness makes sense

Comment (cweitkamp) [reply bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCacheAsync.java]: https://github.com/openhab/openhab-core/pull/1209 - one year later ... :rofl:

Comment (martinvw) [reply bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCacheAsync.java]: 😁
